### PR TITLE
Add hashtags and mentions to Profile, add caption etc to StoryItem

### PR DESCRIPTION
**Pull request by MiguelX413**
Includes https://github.com/instaloader/instaloader/pull/1540.

This pull request adds Profile.biography_hashtags, .biography_mentions, StoryItem.caption, .caption_mentions, .caption_hashtags, and .pcaption.
It also increases code sharing between these and Post.

### CODE REVIEW

**Comment by MiguelX413** [comment]
I could add versionadded information too if you want.

**Comment by aandergr** [comment]
Thanks! This looks like a good change to me. (apart from my observation in #1540).

**Comment by MiguelX413** [comment]
> Thanks! This looks like a good change to me. (apart from my observation in #1540).

How would you recommend I adjust for that and apply that adjustment to all 3 pull requests?

**Comment by aandergr** [comment]
Maybe `r"(?:^|[^\w\n]|_)(?:@)(?!$)(\w(?:(?:\w|(?:\.(?!\.))){0,28}(?:\w))?)"` as a better fitting regex?

**Comment by MiguelX413** [comment]
> Maybe `r"(?:^|[^\w\n]|_)(?:@)(?!$)(\w(?:(?:\w|(?:\.(?!\.))){0,28}(?:\w))?)"` as a better fitting regex?

I think we should be good with https://github.com/instaloader/instaloader/commit/69779ccec372e7307b318f4c053f9fb586f99540 @aandergr

**Comment by MiguelX413** [comment]
I've rebased changes for this PR and #1540 @aandergr

**Comment by MiguelX413** [comment]
https://github.com/instaloader/instaloader/compare/5811afa540cfeed42a56cd3530387ae571d8906d..282807df8f75d30612f3173db884cd25a5b3db90
this would be the change brought about by the rebase.

**Comment by aandergr** [comment]
Could you please rebase this to three commits,
1. one that adds the regex change and refactors the regex out into the more-global namespace
2. one that adds the normalizations
3. the last one that adds the new properties.

**Comment by MiguelX413** [comment]
> Could you please rebase this to three commits,
> 1. one that adds the regex change and refactors the regex out into the more-global namespace
> 2. one that adds the normalizations
> 3. the last one that adds the new properties.

Yes!

**Comment by MiguelX413** [comment]
> Could you please rebase this to three commits,
> 
>     1. one that adds the regex change and refactors the regex out into the more-global namespace
> 
>     2. one that adds the normalizations
> 
>     3. the last one that adds the new properties.

How is it?

File: instaloader/structures.py [hunk 1]
```diff
@@ -42,6 +42,22 @@
...
 PostLocation.lng.__doc__ = "Longitude (:class:`float` or None)."
 
+# This regular expression is by MiguelX413
+_hashtag_regex = re.compile(r"(?:#)((?:\w){1,150})")
+
+# This regular expression is modified from jStassen, adjusted to use Python's \w to
+# support Unicode and a word/beginning of string delimiter at the beginning to ensure
+# that no email addresses join the list of mentions.
+# http://blog.jstassen.com/2016/03/code-regex-for-instagram-username-and-hashtags/
+_mention_regex = re.compile(r"(?:^|[^\w\n]|_])(?:@)(\w(?:(?:\w|(?:\.(?!\.))){0,28}(?:\w))?)", re.ASCII)
```
[the remark this answers is not in the document]
Comment (MiguelX413) [reply instaloader/structures.py:52]: Why don't I just make it `(?:@)(\w(?:(?:\w|(?:\.(?!\.))){0,28}(?:\w))?)`

[the remark this answers is not in the document]
Comment (MiguelX413) [reply instaloader/structures.py]: I think I handled this well now.

**Comment by aandergr** [comment]
Nice, thanks!

**Pull request merged**